### Numerical Analysis
(Formative Assessment 4)

**Romand Lansangan**

---

In [1]:
import numpy as np 
import pandas as pd

## $\int_0^2 x^2 \ln(x^2 + 1) dx$ using $h = 0.25$. 

The interval is $[a, b] = [0, 2]$, and the number of subintervals is $n = \frac{b-a}{h} = \frac{2-0}{0.25} = 8$.

$x_j = a + jh$ for each $j=0,1, . . . , n$ 

In [28]:
n = 8
h_s = np.linspace(0, 2, n+1)
h_s

array([0.  , 0.25, 0.5 , 0.75, 1.  , 1.25, 1.5 , 1.75, 2.  ])

f(x_j)

In [38]:
eq = "(x**2) * ln(x**2 + 1)"

fh = [sip.eq_solver(eq, h) for h in h_s]
fh_ = pd.Series(fh)
fh_

0    0.000000
1    0.003789
2    0.055786
3    0.251036
4    0.693147
5    1.470286
6    2.651974
7    4.293008
8    6.437752
dtype: float64

### Composite Trapezoidal Rule

Recall that the formula for Composite Trapezoidal Rule is:

$$\int_a^b f(x) dx \approx \frac{h}{2} [f(x_0) + 2\sum_{j=1}^{n-1} f(x_j) + f(x_n)]$$




With $h=0.25$ and $n=8$; 

$$
\int_0^2 x^2 \ln(x^2 + 1) dx \approx \frac{0.25}{2} [f(a) + 2\sum_{j=1}^{7} f(x_j) + f(b)]
$$

Lets start with $ 2\sum_{j=1}^{n-1} f(j)$

$$ 2\sum_{j=1}^{7} f(x_j)$$

In [59]:
first_ind = list(range(1,8))

first_ = fh_[first_ind]
first_

1    0.003789
2    0.055786
3    0.251036
4    0.693147
5    1.470286
6    2.651974
7    4.293008
dtype: float64

In [ ]:
first_sum = first_.sum() * 2
first_sum

18.83805374434219

$$
\int_0^2 x^2 \ln(x^2 + 1) dx \approx \frac{0.25}{2} [f(a) + 2\sum_{j=1}^{7} f(x_j) + f(b)]
$$

In [66]:
inside_term = fh_[0] + first_sum + fh_[n]

print("Inside term:", inside_term)

outside_term = 0.25 / 2

approximate_1 = outside_term * inside_term

print("Approximate using Composite Trapezoidal Rule:", approximate_1)

Inside term: 25.275805394078592
Approximate using Composite Trapezoidal Rule: 3.159475674259824


$$\int_0^2 x^2 \ln(x^2 + 1) dx \approx 3.159475674259824$$

### Composite Simpson's Rule
Recall that the formula for composite Simpson's rule is:

$$
\int_a^b f(x) dx = \frac{h}{3} [f(a) + 2 \sum^{(n/2) - 1}_{j=1} f(x_{2j}) + 4 \sum^{n/2}_{j=1} f(x_{2j-1}) + f(b)] - \text{error term} 
$$


With $h=0.25$ and $n=8$; 

$$
\int_0^2 x^2 \ln(x^2 + 1) dx = \frac{0.25}{3} [f(a) + 2 \sum^{3}_{j=1} f(x_{2j}) + 4 \sum^{4}_{j=1} f(x_{2j-1}) + f(b)] - \text{error term} 
$$

For $2 \sum^{(n/2) - 1}_{j=1} f(x_{2j})$

$$2 \sum^{(8/2) - 1}_{j=1} f(x_{2j})$$
$$2 \sum^{3}_{j=1} f(x_{2j})$$
$$2[f(x_2) + f(x_4) + f(x_6)]$$

In [68]:
first_ind = [2, 4, 6]
first_ = fh_[first_ind]*2
first_

2    0.111572
4    1.386294
6    5.303947
dtype: float64

In [69]:
first_sum = first_.sum()
first_sum

6.801813620314403

For $4 \sum^{n/2}_{j=1} f(x_{2j-1})$

$$4 \sum^{4}_{j=1} f(x_{2j-1})$$
$$4[f(x_1) + f(x_3) + f(x_5) + f(x_7)]$$

In [70]:
second_ind = [1,3,5,7]
second_ = fh_[second_ind]*4
second_

1     0.015156
3     1.004146
5     5.881146
7    17.172032
dtype: float64

In [71]:
second_sum = second_.sum()
second_sum

24.072480248055577

$$
\int_0^2 x^2 \ln(x^2 + 1) dx = \frac{0.25}{3} [f(a) + 2 \sum^{3}_{j=1} f(x_{2j}) + 4 \sum^{4}_{j=1} f(x_{2j-1}) + f(b)] - \text{error term} 
$$

In [73]:
fa = fh_[0]
fb = fh_[n]
inside_term = fa+fb+first_sum+second_sum
print("Inside Term:", inside_term)

outside_term = 0.25 / 3

approximate_2 = inside_term*outside_term
print("Approximate using Composite Simpson's Rule:", approximate_2)

Inside Term: 37.31204551810638
Approximate using Composite Simpson's Rule: 3.1093371265088647


$$\int_0^2 x^2 \ln(x^2 + 1) dx \approx 3.1093371265088647$$

### Composite Midpoint Rule

Recall that the formula for Composite Midpoint Rule is:

$$
\int_a^b f(x) dx \approx 2h \sum^{n/2}_{j=0} f(x_{2j})
$$

but h is different in this.

$$h = \frac{b-a}{n+2} $$

$$0.25 = \frac{2-0}{n+2}$$

$$n+2 = 2/0.25 = 8$$

$$ n = 8-2 = 6$$

$x_j =a + (j+1)h$ for $j = -1, 0, ..., n+1$
 

In [85]:
n = 6
h = 0.25
a = 0
x = [a + (j+1)*h for j in range(-1, n+2)]
f_x = [sip.eq_solver(eq, x) for x in x]
f_x_ = pd.Series(f_x, index=list(range(-1,n+2)))

f_x_

-1    0.000000
 0    0.003789
 1    0.055786
 2    0.251036
 3    0.693147
 4    1.470286
 5    2.651974
 6    4.293008
 7    6.437752
dtype: float64



$$\int_0^2 x^2 \ln(x^2 + 1) dx \approx 2(0.25) \sum^{3.5}_{j=0} f(x_{2j})$$


$$ \approx 0.5[f(0) + f(2) + f(4) + f(6)]$$


In [88]:
f_ind = list(range(0,7, 2))

inside_term = f_x_[f_ind]
inside_term

0    0.003789
2    0.251036
4    1.470286
6    4.293008
dtype: float64

In [90]:
inside_term_sum = inside_term.sum()

print("Inside Sum:", inside_term_sum)

approximate_3 = 0.5 * inside_term_sum

print("Approximate using Composite Midpoint Rule:", approximate_3)

Inside Sum: 6.018120062013894
Approximate using Composite Midpoint Rule: 3.009060031006947


$$\int_0^2 x^2 \ln(x^2 + 1) dx \approx 3.009060031006947$$

---


## Determine the values of n and h required to approximate $\int_0^2 \frac{1}{x+4} dx$ to within $10^{-5}$

### Composite Trapezoidal Rule

Recall that the error term for this rule is:
$$\frac{b-a}{12}h^2 f'' (\mu)$$

where $\mu \in (a,b)$

$$f(x) =  \frac{1}{x+4}$$

$$f'(x) = -\frac{1}{(x+4)^2}$$

$$f''(x) = \frac{2}{(x+4)^3}$$

Since $\mu \in (0,2)$,

$$|\frac{1}{6} h^2 f''(\mu)| < \frac{1}{6} h^2 f(0) = \frac{1}{6} h^2 \frac{2}{4^3}$$

$$|\frac{1}{6} h^2 f''(\mu)| < \frac{1}{6} h^2 \frac{2}{64} = \frac{2}{384} h^2$$


Thus, solving for 

$$ \frac{2}{384} h^2 < 10^{-5}$$

$$h^2 < \frac{384}{10^5 \times 2}$$

$$h < \sqrt{\frac{384}{10^5 \times 2}}$$


In [97]:
(384/((10**5) * 2))**0.5

0.04381780460041329

$$h < 0.04381780460041329$$

Recall that

$$h = \frac{b-a}{n}$$

$$h = \frac{2}{n} < 0.04381780460041329$$

$$n > \frac{2}{0.04381780460041329}$$

In [98]:
2/0.04381780460041329

45.64354645876384

$$n > 45.64354645876384$$
$$n \geq 46$$

---

### Composite Simpson's Rule

Recall that the error term for this rule is:

$$\frac{b-a}{180}h^4 f^{(4)}(\mu)$$

where $\mu \in (a,b)$

$$f^{(3)}(x) = \frac{-6}{(x+4)^4}$$

$$f^{(4)}(x) = \frac{24}{(x+4)^5}$$

$$|\frac{2}{180}h^4 f^{(4)}(\mu)| < \frac{2}{180}h^4 \frac{24}{4^6}$$

$$|\frac{2}{180}h^4 f^{(4)}(\mu)| < \frac{2}{180}h^4 \frac{24}{4096}$$

$$|\frac{2}{180}h^4 f^{(4)}(\mu)| < \frac{48}{737280} h^4$$

Solving for:

$$\frac{48}{737280} h^4 < \frac{1}{10^5}$$

$$h^4 < \frac{737280}{48 \times 10^5}$$

$$h < (\frac{737280}{48 \times 10^5})^{1/4}$$

In [ ]:
(737280/(48 * (10**5))) ** (1/4)

0.626033832029315

$$h < 0.626033832029315$$

$$h = \frac{b-a}{n} = \frac{2}{n} < 0.626033832029315$$

$$n > \frac{2}{0.626033832029315}$$

In [105]:
2/0.626033832029315

3.194715521231362

$$n>3.194715521231362$$

$$n \geq 4$$

---

### Composite Midpoint Rule

Recall that this rule has an error term of:

$$\frac{b-a}{6}h^2 f''(\mu)$$

where $\mu \in (a,b)$

$$|\frac{2}{6} h^2 f''(\mu)| < \frac{1}{3}h^2 \frac{2}{64} = \frac{2}{192}h^2$$

Solving for:

$$\frac{2}{192}h^2 < \frac{1}{10^5}$$

$$h^2 < \frac{192}{2 \times 10^5}$$

$$h < \sqrt{\frac{192}{2 \times 10^5}}$$

In [109]:
(192 / (2 * (10**5)))**0.5

0.030983866769659335

$$h < 0.030983866769659335$$

Recall that:
$$h=\frac{b-a}{n+2}$$

$$h = \frac{2}{n+2} < 0.030983866769659335$$

$$n+2 > \frac{2}{0.030983866769659335}$$

$$n > \frac{2}{0.030983866769659335} - 2$$

In [110]:
(2/0.030983866769659335) - 2

62.549722436790276

$$n > 62.549722436790276$$

$$n \geq 63$$

---

Since the Guassian quarature rules are defined on the interval $[-1,1]$, we ought to transpose the interval of integration. 

$$x(t) = \frac{(b-a)t + (b+a)}{2} = \frac{(1-0)t + (1+0)}{2} = \frac{t+1}{2}$$
    
Then, $dx = \frac{b-a}{2} dt = \frac{1}{2} dt$.

Consquentyly, $f(x) = x^2 e^x$ becomes $f\left(\frac{t+1}{2}\right) = \left(\frac{t+1}{2}\right)^2 e^{\frac{t+1}{2}}$.

$$\int_0^1 x^2 e^x dx = \int_{-1}^1 f\left(\frac{t+1}{2}\right) \frac{1}{2} dt = \int_{-1}^1 \frac{1}{2} \left(\frac{t+1}{2}\right)^2 e^{\frac{t+1}{2}} dt$$

Let $g(t) = \frac{1}{2} \left(\frac{t+1}{2}\right)^2 e^{\frac{t+1}{2}}$.

Let us now find the nodes and the wieghts for $n=2$.

From the slide (Page 36), the nodes ($t_i$) and weights ($c_i$) needed for Guassian Quarature are:

$t_1 = -0.5773502692$, $c_1 = 1.0$

$t_2 = 0.5773502692$, $c_2 = 1.0$

Recall that the formula for Guassian Quarature rule is:

$$\int_{-1}^1 g(t) dt \approx \sum_{i=1}^n c_i g(t_i)$$

For $n=2$:
$$\int_{-1}^1 g(t) dt \approx c_1 g(t_1) + c_2 g(t_2)$$

In [10]:
eq = "0.5*(((x+1)/2)**2 )* e^(x/2) * e^(0.5)"
ts = [-0.5773502692, 0.5773502692]

gt = [sip.eq_solver(eq, t) for t in ts]
gt

[0.027583440137066442, 0.6843583341162582]

In [11]:
print("Approximation:", sum(gt))

Approximation: 0.7119417742533247


$$\int_{-1}^1 g(t) dt \approx \sum_{i=1}^n c_i g(t_i) \approx 0.7119417742533247$$

---

In [13]:
import numpy as np
import math 

def gaussian_double_integral(a, b, c_func, d_func, f_func, x_nodes, x_weights, y_nodes, y_weights):
    """
    Approximates the double integral int_a^b int_c(x)^d(x) f(x, y) dy dx
    using iterated Gaussian quadrature.

    Args:
        a (float): Lower limit of the outer integral (x).
        b (float): Upper limit of the outer integral (x).
        c_func (function): Function c(x) for the lower limit of the inner integral.
        d_func (function): Function d(x) for the upper limit of the inner integral.
        f_func (function): The integrand function f(x, y).
        x_nodes (list or array): Gaussian quadrature nodes for the outer (x) integral on [-1, 1].
        x_weights (list or array): Gaussian quadrature weights for the outer (x) integral.
        y_nodes (list or array): Gaussian quadrature nodes for the inner (y) integral on [-1, 1].
        y_weights (list or array): Gaussian quadrature weights for the inner (y) integral.

    Returns:
        float: The approximate value of the double integral.
    """
    # Determine number of points for x and y integration
    m = len(x_nodes)
    n = len(y_nodes)

    # Step 1: Initialize h1, h2, J
    h1 = (b - a) / 2.0  # Transformation factor for x interval
    h2 = (b + a) / 2.0  # Shift factor for x interval
    J = 0.0             # Initialize the total approximation

    # Step 2: Outer loop (over x nodes)
    # The pseudocode uses 1-based indexing (i=1 to m).
    # In Python, we use 0-based indexing (i=0 to m-1).
    for i in range(m):
        # Step 3: Transformations for the outer node
        # Transform the outer node (x_nodes[i]) from [-1, 1] to [a, b]
        x = h1 * x_nodes[i] + h2

        # Get the inner integral limits at this x
        c1 = c_func(x)
        d1 = d_func(x)

        # Step 3: Transformations for the inner integral interval [c1, d1]
        k1 = (d1 - c1) / 2.0 # Transformation factor for y interval
        k2 = (d1 + c1) / 2.0 # Shift factor for y interval

        # Step 3/4: Initialize inner sum variable
        JX = 0.0 # Initialize the sum for the inner integral approximation

        # Step 4: Inner loop (over y nodes)
        # Loop through the nodes/weights for the inner integral (n points).
        # The pseudocode uses 1-based indexing (j=1 to n).
        # In Python, we use 0-based indexing (j=0 to n-1).
        for j in range(n):
            # Step 5: Transformation and function evaluation for the inner node
            # Transform the inner node (y_nodes[j]) from [-1, 1] to [c1, d1]
            y = k1 * y_nodes[j] + k2

            # Evaluate the integrand at (x, y)
            Q = f_func(x, y)

            # Step 5: Accumulate inner weighted sum
            # JX approximates the sum part of the inner integral approximation: sum(w_j * f(x_i, y_j))
            JX = JX + y_weights[j] * Q

        # Step 6: Accumulate outer weighted sum
        # JX now holds sum(w_j * f(x_i, y_j)) for the fixed x_i.
        # (k1 * JX) is the approximation of the inner integral at x_i.
        # x_weights[i] is the weight for the outer integral at x_i.
        J = J + x_weights[i] * k1 * JX

    # Step 7: Final scaling by the outer integral transformation factor (b-a)/2 = h1
    J = h1 * J

    # Step 7: Output (return) the result
    return J

# --- How to use the function ---
# You need to provide the Gaussian quadrature nodes and weights for the
# desired number of points (m and n). These are standard values for the
# interval [-1, 1] and can be found in tables or generated by libraries.
# For example, using numpy:
# from numpy.polynomial.legendre import leggauss
# nodes_m, weights_m = leggauss(m)
# nodes_n, weights_n = leggauss(n)

# Example Gaussian Quadrature Nodes and Weights (from Page 36, rounded)
# For n=4 (as required by the problem)
nodes_4 = np.array([-0.8611363116, -0.3399810436, 0.3399810436, 0.8611363116])
weights_4 = np.array([0.3478548451, 0.6521451549, 0.6521451549, 0.3478548451])

# Define the specific functions for problem 4:
# f(x, y) = x^2 + sqrt(y)
# c(x) = 0
# d(x) = x
# Outer limits: a=1, b=1.5
# m=4, n=4

def c_func_q4(x):
    return 0.0

def d_func_q4(x):
    return x

def f_func_q4(x, y):
    # Handle potential sqrt of negative number if transformation is incorrect,
    # but for this problem y should be >= 0 in the integration domain.
    return x**2 + math.sqrt(y)

# Example usage for problem 4 with m=4 and n=4:
Approximation = gaussian_double_integral(1.0, 1.5, c_func_q4, d_func_q4, f_func_q4, nodes_4, weights_4, nodes_4, weights_4)
print(f"Approximation for the double integral: {Approximation}")

Approximation for the double integral: 1.48462057710937


In [2]:
import math
import regex
import pandas as pd

class Sipnayan:
    def __init__(self, math_object, regex, pd):
        self.math_object = math_object
        self.reg = regex
        self.pd = pd

    def number_solver(self, equation):
        return eval(equation)
    
    def eq_solver(self, equation, var_val, var="x"):
        """"var != e
        Note to future romand: make varaible a list instead for equations beyond 2d
        """
        # print(f"Original: {equation}")
        equation = equation.replace(var, str(var_val))
        # print(f"Parsed: {equation}")
        # Check for other variable other than specified var
        if (self.check_for_other_var(equation, var)):
            return f"Letters detected aside from independent variable ({var})"
        operations = ["e", "cos", "sin", "tan", "ln"]
        if any(operation in equation for operation in operations):
            return eval(self.nested_handler(equation))
        return self.number_solver(equation)

    def trigo(self, trig_op, arg):
        match str(trig_op):
            case "cos":
                return self.math_object.cos(arg)
            case "sin":
                return self.math_object.sin(arg)
            case "tan":
                return self.math_object.tan(arg)
            case default:
                return "invalid argument"
                
    def exp_solve(self, arg):
        """For e^x with x as arg"""
        return self.math_object.exp(arg)

    def ln_solve(self, arg):
        return self.math_object.log(arg)
    
    def nested_handler(self, equation):
        operations = ["e\\^", "cos", "sin", "tan", "ln"]
        ops = ["e", "cos", "sin", "tan", "ln"]
        # print(f"Processing equation: {equation}")
        
        pat = rf'({"|".join(operations)})\(((?:[^\(\)]+|(?R))*)\)'
        # print(f"Regex pattern: {pat}")
        
        while any(operation in equation for operation in ops):
            match = regex.search(pat, equation)
            if not match:
                break

            opp = match.group(1)
            ovr_expression = match.group(0)  
            argument = match.group(2) 
            # print(f"Matched operation: {ovr_expression}, Argument: {argument}")
    
            if any(operation in argument for operation in ops):
                argument = self.nested_handler(argument)
    
            equation = self.special_operations(opp, ovr_expression, argument, equation)
            # print(f"Updated equation: {equation}")
        
        return equation
            
    def special_operations(self, opp, ovr_expression, argument, equation):
        # print(f"Processing {opp} with argument: {argument}")
        try:
            if "e" in opp:
                result = self.exp_solve(eval(argument))
            elif "ln" in opp:
                result = self.ln_solve(eval(argument))
            else:
                result = self.trigo(opp, eval(argument))
        except Exception as e:
            print(f"Error in {opp}: {e}")
            return equation
    
        updated_equation = equation.replace(ovr_expression, str(result))
        return updated_equation



    def check_for_other_var(self, equation, var):
        remove = ['cos', 'sin', 'tan', 'e', "ln"]
        for opp in remove:
            equation = equation.replace(opp,"")
        for i in equation:
            if i.isalpha() and i != var:
                return True
        return False

sip = Sipnayan(math, regex, pd)